### ISSUE TITLE
Support getter setters in interfaces

### ISSUE THREAD

**Issue by mhevery** [Microsoft/TypeScript#3745]
Classes support getters / setters

```
class MyClass {
  get value(): string { return 'value'; }
  set value(v: string): string {}
}
```

but the corresponding interface is not supported:

```
interface MyInterface {
  get value(): string;
  set value(v: string): string;
}
```

**Comment by RyanCavanaugh** [comment]
You simply declare it as a property.
```ts
interface MyInterface {
  value: string;
}
```
Whether a property is implemented as a field or a getter/setter pair is an implementation detail, not part of the type.

**Comment by mhevery** [comment]
Sorry for not being more clear. How do I create an interface which only has getters (no setters). Please reopen.

```
interface IName {
  get name():string; // SYNTAX ERROR
}

class User implements IName {
  name: string; // SHOULD WORKS
}

class User2 implements IName {
  get name() { return value; } // ALSO SHOULD WORKS
}

class User2 implements IName {
  set name(value) { } // ALSO SHOULD FAIL
}
```

**Comment by jonathandturner** [comment]
We've had an outstanding feature request for read-only: https://github.com/Microsoft/TypeScript/issues/12  You might want to pile on there, also.

**Comment by alexeagle** [comment]
Now that we have `abstract`, I think this problem is solved:
```
abstract class HasReadonlyProp {
  abstract get foo(): number;
}
class HasReadonlyPropImpl extends HasReadonlyProp {
  get foo(): number { return 1; }
}
```
```
function ClientCode() {
  let h = new HasReadonlyProp();
  console.log(h.foo); //=> 1
  h.foo = 2; //=> typechecking error 
}
```

**Comment by GameDaddy** [comment]
abstract is a part of single class inheritance, this does not solve the problem when you want an interface definition added to an existing class hierarchy.

**Comment by kitsonk** [comment]
This is really solved now fully with `readonly` which signifies and interface where there is only a getter (or an interface where you logically want to disallow assignment at design time).

```typescript
interface IName {
  readonly name(): string; // not a syntax error
}
```

**Comment by hillin** [comment]
@kitsonk  yes, but only without parenthesizes:
```typescript
interface IName {
  readonly name:: string; // not a syntax error
}
```

**Comment by xtravar** [comment]
This feature would be useful for TypeScript Java interop declarations.  I'm compiling TS into JS for use with Java's Nashorn scripting engine, and have been declaring the Java classes/interfaces in TS for better type safety and code completion.

Java does not natively support properties.  However, Nashorn will pass-through property usage to an object's appropriately named getter and setter.

All this is to say, copying Java declarations into TypeScript would be significantly easier and with less work, because the Java interface already has a getter and setter method, with full documentation.  Converting this to the other property syntax is cumbersome.

I totally realize it's a weird edge case and probably shouldn't influence TypeScript's development, but the lack of this syntax does seem a little inconsistent.

**Comment by aluanhaddad** [comment]
That's kind of ironic since Java still doesn't have properties. It sounds like your use case would be best addressed by code generation.

**Comment by amir-arad** [comment]
I'm writing a container that should inject an `address` field to its elements. The elements can currently be any `object`.
I want to be able to declare the type of the elements as one that may or may not have a value under `address`, but supports setting the `address` field to string values only.

a setter would be a nice fit IMO:
```ts
interface Element {
  set address(a:string):void;
}
```

**Comment by andrey-zakharov** [comment]
I wish we could just type:
```
interface ISome {
 field: string;
}
class C implements ISome {
 public get field() { return _field; }
 public set field(v: string) { // MY STUFF; _field = field; }
}
```
and it should works in TS, maybe with some warnings about missing overrides keyword :)

IMHO.....IMHO.....IMHO.....IMHO.....IMHO.....IMHO.....IMHO.....IMHO.....

**Comment by tncbbthositg** [comment]
@RyanCavanaugh would you be willing to weight in on my latest comment?  I think React functional components have created an example where the ability of the TypeScript interface to describe the surface of an object has exceeded the ability of the developer to create that object without instructing the compiler to ignore certain errors.

Thank you

**Comment by RyanCavanaugh** [comment]
> What I really want is to require a getter to be implemented on the type (or a readonly value to be set on it . . . I don't really care), but then to be able to set that property without having to tell the compiler I know I'm doing something it thinks is wrong but that it's not wrong.

I don't think the sort of temporary-mutability you're describing here rises to the level of something that must have a solution. You can solve this with various safe upcasts and/or helper functions today.

**Comment by tncbbthositg** [comment]
Thanks @RyanCavanaugh!  I really appreciate you taking the time to look at this again!

I was going through the attempts I had originally made to satisfy the original interface in a type safe way without type errors.  Here's the original interface again for context:

```typescript
interface GreeterFunction {
    (name: string): string;
    readonly greeting: string;
}
```

This behaves correctly, but the type checker doesn't like setting the read-only property:

```typescript
const hiyoGreeter: GreeterFunction = (name) => `${hiyoGreeter.greeting} ${name}`;
hiyoGreeter.greeting = 'Hiyo';
```

This one works correctly, but the type checker complains that `property 'greeting' is missing in type`:

```typescript
type Mutable<T> = {
  -readonly[P in keyof T]: T[P]
};

const hiyoGreeter: GreeterFunction = (name) => `${hiyoGreeter.greeting} ${name}`;
(hiyoGreeter as Mutable<GreeterFunction>).greeting = 'Hiyo';
```

This works, but again `property 'greeting' is missing in type`:

```typescript
const hiyoGreeter: GreeterFunction = (name) => `${hiyoGreeter.greeting} ${name}`;
Object.defineProperty(hiyoGreeter, 'greeting', { value: 'Hiyo' });
```

This works and the type checker doesn't complain:

```typescript
const hiyoGreeter: GreeterFunction = ((name) => `${hiyoGreeter.greeting} ${name}`) as GreeterFunction;
Object.defineProperty(hiyoGreeter, 'greeting', { value: 'Hiyo' });
```

But, it loses the value of the type system because this doesn't work and the type checker still doesn't complain:

```typescript
const hiyoGreeter: GreeterFunction = ((name) => `${hiyoGreeter.greeting} ${name}`) as GreeterFunction;
// Object.defineProperty(hiyoGreeter, 'greeting', { value: 'Hiyo' });
```

Basically, I can tell the type checker to "trust me," but I don't want the type checker to trust me; I want to trust the type checker.  

I'm sure I'm missing something obvious, but I don't see how I can define an object that meets that interface without either ignoring an error or making a type assertion that exposes a risk of runtime errors instead of compile time errors.  Could you potentially point me in the right direction?

Thanks again!!
